File: 5_prj/mtk/m3308/hardware/fingerprint/fpc1552s/fingerprint_extension/api/src/main/java/com/fingerprints/extension/authenticator2/FpcFingerprintAuthenticator.java
/**
 * Copyright (c) 2021 Fingerprint Cards AB <tech@fingerprints.com>
 * All rights are reserved.
 * Proprietary and confidential.
 * Unauthorized copying of this file, via any medium is strictly prohibited.
 * Any use is subject to an appropriate license granted by Fingerprint Cards AB.
 */

package com.fingerprints.extension.authenticator2;

import android.os.RemoteException;

import java.util.ArrayList;

import com.fingerprints.extension.V1_0.IFpcFingerprintAuthenticator;
import com.fingerprints.extension.V1_0.IAuthenticatorCallback;
import com.fingerprints.extension.util.Logger;
import com.fingerprints.extension.util.ArrayUtils;

public class FpcFingerprintAuthenticator {
    private Logger mLogger = new Logger(getClass().getSimpleName());
    private final static String LOG_TAG = FpcFingerprintAuthenticator.class.getSimpleName();

    private IFpcFingerprintAuthenticator mService;
    private IAuthenticatorServiceReceiver mReceiver;

    public interface IAuthenticatorServiceReceiver {
        void onEnrollResult(int fingerId, int remaining);
        void onAuthenticated(int fingerId);
        void onEnumerate(int fingerId, int remaining);
        void onRemoved(int fingerId, int remaining);
        void onAcquired(int status);
        void onError(int errorCode);
    }

    private final IAuthenticatorCallback mCallback = new IAuthenticatorCallback.Stub() {
        @Override
        public void onEnrollResult(int fingerId, int remaining) {
            mLogger.d(LOG_TAG + "onEnrollResult");
            mReceiver.onEnrollResult(fingerId, remaining);
        }

        @Override
        public void onAuthenticated(int fingerId) {
            mLogger.d(LOG_TAG + "onAuthenticated");
            mReceiver.onAuthenticated(fingerId);
        }

        @Override
        public void onEnumerate(int fingerId, int remaining) {
            mLogger.d(LOG_TAG + "onEnumerate");
            mReceiver.onEnumerate(fingerId, remaining);
        }

        @Override
        public void onRemoved(int fingerId, int remaining) {
            mLogger.d(LOG_TAG + "onRemoved");
            mReceiver.onRemoved(fingerId, remaining);
        }

        @Override
        public void onAcquired(int status) {
            mLogger.d(LOG_TAG + "onAcquired");
            mReceiver.onAcquired(status);
        }

        @Override
        public void onError(int errorCode) {
            mLogger.d(LOG_TAG + "onError");
            mReceiver.onError(errorCode);
        }
    };

    public FpcFingerprintAuthenticator() throws RemoteException {
        mLogger.d(LOG_TAG + "FpcFingerprintAuthenticator");

        mService = IFpcFingerprintAuthenticator.getService();
        if (mService == null) {
            throw new RemoteException("Failed to get FpcFingerprintAuthenticator service");
        } else {
            try {
                mService.setNotify(mCallback);
            } catch (RemoteException e) {
                mLogger.e(LOG_TAG + "RemoteException: ", e);
            }
        }

        mLogger.d(LOG_TAG + "FpcFingerprintAuthenticator");
    }

    public void enroll(IAuthenticatorServiceReceiver receiver) {
        mLogger.d(LOG_TAG + "enroll");

        mReceiver = receiver;

        if (mService != null) {
            try {
                mService.enroll();
            } catch (RemoteException e) {
                mLogger.e(LOG_TAG + "RemoteException: ", e);
            }
        }

        mLogger.d(LOG_TAG + "enroll");
    }

    public void authenticate(IAuthenticatorServiceReceiver receiver) {
        mLogger.d(LOG_TAG + "authenticate");

        mReceiver = receiver;

        if (mService != null) {
            try {
                mService.authenticate();
            } catch (RemoteException e) {
                mLogger.e(LOG_TAG + "RemoteException: ", e);
            }
        }

        mLogger.d(LOG_TAG + "authenticate");
    }

    public void enumerate(IAuthenticatorServiceReceiver receiver) {
        mLogger.d(LOG_TAG + "enumerate");

        mReceiver = receiver;

        if (mService != null) {
            try {
                mService.enumerate();
            } catch (RemoteException e) {
                mLogger.e(LOG_TAG + "RemoteException: ", e);
            }
        }

        mLogger.d(LOG_TAG + "enumerate");
    }

    public void remove(int id, IAuthenticatorServiceReceiver receiver) {
        mLogger.d(LOG_TAG + "remove");

        mReceiver = receiver;

        if (mService != null) {
            try {
                mService.remove(id);
            } catch (RemoteException e) {
                mLogger.e(LOG_TAG + "RemoteException: ", e);
            }
        }

        mLogger.d(LOG_TAG + "remove");
    }

    public void cancel() {
        mLogger.d(LOG_TAG + "cancel");

        if (mService != null) {
            try {
                mService.cancel();
            } catch (RemoteException e) {
                mLogger.e(LOG_TAG + "RemoteException: ", e);
            }
        }

        mLogger.d(LOG_TAG + "cancel");
    }
}

File: 5_prj/mtk/m3308/hardware/fingerprint/fpc1552s/fingerprint_hal/fingerprint_hwbinder/extensions/authenticator2/FpcFingerprintAuthenticator.cpp
/**
 * Copyright (c) 2021 Fingerprint Cards AB <tech@fingerprints.com>
 * All rights are reserved.
 * Proprietary and confidential.
 * Unauthorized copying of this file, via any medium is strictly prohibited.
 * Any use is subject to an appropriate license granted by Fingerprint Cards AB.
 */

#include <utils/Log.h>

#include "FpcFingerprintAuthenticator.h"

void add_authenticator_service_2(fpc_authenticator_2_t* device) {
    ALOGE("fpc fingerprint %s", __func__);
    com::fingerprints::extension::V1_0::implementation::FpcFingerprintAuthenticator::instantiate(
            device);
}

namespace com {
namespace fingerprints {
namespace extension {
namespace V1_0 {
namespace implementation {

FpcFingerprintAuthenticator* FpcFingerprintAuthenticator::sInstance = NULL;

void FpcFingerprintAuthenticator::instantiate(fpc_authenticator_2_t* device) {
    if (sInstance == NULL) {
        sInstance = new FpcFingerprintAuthenticator(device);
        if (sInstance->registerAsService() != android::OK) {
            ALOGE("Failed to register FpcFingerprintAuthenticator");
        }
    }
}

FpcFingerprintAuthenticator::FpcFingerprintAuthenticator(fpc_authenticator_2_t* device)
        : mCallback(NULL),
          mDevice(device) {
    mCompatCallback.on_enroll_result = onEnrollResult;
    mCompatCallback.on_authenticated = onAuthenticated;
    mCompatCallback.on_enumerate = onEnumerate;
    mCompatCallback.on_removed = onRemoved;
    mCompatCallback.on_acquired = onAcquired;
    mCompatCallback.on_error = onError;

    if (mDevice) {
        mDevice->init(mDevice, &mCompatCallback, this);
    } else {
        ALOGE("%s: device not yet registered", __func__);
    }
}

void FpcFingerprintAuthenticator::onEnrollResult(void *context, uint32_t id, uint32_t remaining) {
    FpcFingerprintAuthenticator * self = static_cast<FpcFingerprintAuthenticator*>(context);
    if (self->mCallback == NULL) {
        ALOGE("%s: callback not yet registered", __func__);
        return;
    }

    if (!self->mCallback->onEnrollResult(id, remaining).isOk()) {
        ALOGE("%s: call onEnrollResult failed", __func__);
    }
}

void FpcFingerprintAuthenticator::onAuthenticated(void *context, uint32_t id) {
    FpcFingerprintAuthenticator * self = static_cast<FpcFingerprintAuthenticator*>(context);
    if (self->mCallback == NULL) {
        ALOGE("%s: callback not yet registered", __func__);
        return;
    }

    if (!self->mCallback->onAuthenticated(id).isOk()) {
        ALOGE("%s: call onAuthenticated failed", __func__);
    }
}

void FpcFingerprintAuthenticator::onEnumerate(void *context, uint32_t id, uint32_t remaining) {
    FpcFingerprintAuthenticator * self = static_cast<FpcFingerprintAuthenticator*>(context);
    if (self->mCallback == NULL) {
        ALOGE("%s: callback not yet registered", __func__);
        return;
    }

    if (!self->mCallback->onEnumerate(id, remaining).isOk()) {
        ALOGE("%s: call onEnumerate failed", __func__);
    }
}

void FpcFingerprintAuthenticator::onRemoved(void *context, uint32_t id, uint32_t remaining) {
    FpcFingerprintAuthenticator * self = static_cast<FpcFingerprintAuthenticator*>(context);
    if (self->mCallback == NULL) {
        ALOGE("%s: callback not yet registered", __func__);
        return;
    }

    if (!self->mCallback->onRemoved(id, remaining).isOk()) {
        ALOGE("%s: call onRemoved failed", __func__);
    }
}

void FpcFingerprintAuthenticator::onAcquired(void* context, int32_t code) {
    FpcFingerprintAuthenticator *self = static_cast<FpcFingerprintAuthenticator*>(context);
    if (self->mCallback == NULL) {
        ALOGE("%s: callback not yet registered", __func__);
        return;
    }

    if (!self->mCallback->onAcquired(code).isOk()) {
        ALOGE("%s: call onAcquired failed", __func__);
    }
}

void FpcFingerprintAuthenticator::onError(void* context, int32_t code) {
    FpcFingerprintAuthenticator *self = static_cast<FpcFingerprintAuthenticator*>(context);
    if (self->mCallback == NULL) {
        ALOGE("%s: callback not yet registered", __func__);
        return;
    }

    if (!self->mCallback->onError(code).isOk()) {
        ALOGE("%s: call onError failed", __func__);
    }
}

Return<void> FpcFingerprintAuthenticator::setNotify(const sp<IAuthenticatorCallback>& callback)
{
    mCallback = callback;
    return Void();
}

Return<void> FpcFingerprintAuthenticator::enroll() {
    if (mDevice) {
        mDevice->enroll(mDevice);
    } else {
        ALOGE("%s: device not yet registered", __func__);
    }
    return Void();
}

Return<void> FpcFingerprintAuthenticator::authenticate() {
    if (mDevice) {
        mDevice->authenticate(mDevice);
    } else {
        ALOGE("%s: device not yet registered", __func__);
    }
    return Void();
}

Return<void> FpcFingerprintAuthenticator::enumerate() {
    if (mDevice) {
        mDevice->enumerate(mDevice);
    } else {
        ALOGE("%s: device not yet registered", __func__);
    }
    return Void();
}

Return<void> FpcFingerprintAuthenticator:: remove(uint32_t id) {
    if (mDevice) {
        mDevice->remove(mDevice, id);
    } else {
        ALOGE("%s: device not yet registered", __func__);
    }
    return Void();
}

Return<void> FpcFingerprintAuthenticator::cancel() {
    if (mDevice) {
        mDevice->cancel(mDevice);
    } else {
        ALOGE("%s: device not yet registered", __func__);
    }
    return Void();
}

}  // namespace implementation
}  // namespace V1_0
}  // namespace extension
}  // namespace fingerprints
}  // namespace com
